kv/memberlist/kv_init_service.go

```diff
@@ -95,191 +86,5 @@ func (kvs *KVInitService) stopping(_ error) error {
-
-	v := pageData{
-		Now:              time.Now(),
-		Memberlist:       kv.memberlist,
-		SortedMembers:    members,
-		Store:            kv.storeCopy(),
-		SentMessages:     sent,
-		ReceivedMessages: received,
-	}
-
-	accept := req.Header.Get("Accept")
-	if strings.Contains(accept, "application/json") {
-		w.Header().Set("Content-Type", "application/json")
-
-		if err := json.NewEncoder(w).Encode(v); err != nil {
-			http.Error(w, err.Error(), http.StatusInternalServerError)
-		}
-		return
-	}
-
-	w.Header().Set("Content-Type", "text/html")
-	if err := defaultPageTemplate.Execute(w, v); err != nil {
-		http.Error(w, err.Error(), http.StatusInternalServerError)
-	}
-}
-
-func getFormat(req *http.Request) string {
-	const viewFormat = "format"
-
-	format := ""
-	if len(req.Form[viewFormat]) > 0 {
-		format = req.Form[viewFormat][0]
-	}
-	return format
-}
-
-func viewMessage(w http.ResponseWriter, kv *KV, msg message, format string) {
-	c := kv.GetCodec(msg.Pair.Codec)
-	if c == nil {
-		http.Error(w, "codec not found", http.StatusNotFound)
-		return
-	}
-
-	val, err := c.Decode(msg.Pair.Value)
-	if err != nil {
-		http.Error(w, fmt.Sprintf("failed to decode: %v", err), http.StatusInternalServerError)
-		return
-	}
-
-	formatValue(w, val, format)
-}
-
-func viewKey(w http.ResponseWriter, store map[string]valueDesc, key string, format string) {
-	if store[key].value == nil {
-		http.Error(w, "value not found", http.StatusNotFound)
-		return
-	}
-
-	formatValue(w, store[key].value, format)
-}
-
-func formatValue(w http.ResponseWriter, val interface{}, format string) {
-	w.WriteHeader(200)
-	w.Header().Add("content-type", "text/plain")
-
-	switch format {
-	case "json", "json-pretty":
-		enc := json.NewEncoder(w)
-		if format == "json-pretty" {
-			enc.SetIndent("", "    ")
-		}
-
-		err := enc.Encode(val)
-		if err != nil {
-			http.Error(w, err.Error(), http.StatusInternalServerError)
-		}
-
-	default:
-		_, _ = fmt.Fprintf(w, "%#v", val)
-	}
-}
-
-func downloadKey(w http.ResponseWriter, kv *KV, store map[string]valueDesc, key string) {
-	if store[key].value == nil {
-		http.Error(w, "value not found", http.StatusNotFound)
-		return
-	}
-
-	val := store[key]
-
-	c := kv.GetCodec(store[key].codecID)
-	if c == nil {
-		http.Error(w, "codec not found", http.StatusNotFound)
-		return
-	}
-
-	encoded, err := c.Encode(val.value)
-	if err != nil {
-		http.Error(w, fmt.Sprintf("failed to encode: %v", err), http.StatusInternalServerError)
-		return
-	}
-
-	w.Header().Add("content-type", "application/octet-stream")
-	// Set content-length so that client knows whether it has received full response or not.
-	w.Header().Add("content-length", strconv.Itoa(len(encoded)))
-	w.Header().Add("content-disposition", fmt.Sprintf("attachment; filename=%d-%s", val.version, key))
-	w.WriteHeader(200)
-
-	// Ignore errors, we cannot do anything about them.
-	_, _ = w.Write(encoded)
+	NewHTTPStatusHandler(kvs, defaultPageTemplate).ServeHTTP(w, req)
```

Comment:
[nit] Instead of creating an handler for each request, what if we create it once and just reuse it?

Reply:
Same pattern is followed in other places (ring status handler) and it doesn't really cause an overhead (just an interface allocation, and even if it did cause an overhead... this is an http handler, so the impact is neglible). 

OTOH, having an instantiated handler in the kvs would mean that there's a field in kvs that refers to kvs, which is something I dislike.

Since there's no hard reason to have the field, I'd prefer keeping it this way.

Reply:
Sure!